test/QuerySpec.hs

```diff
@@ -18,55 +21,85 @@ testDBName = "mongodb-haskell-test"
-                              , ["name" =: "Indians", "league" =: "American"]
-                              ]
+  _ <-
+    db $
+      testInsert
+        "team"
+        [ ["name" =: "Yankees", "league" =: "American"]
+        , -- Try to insert document with duplicate key
+          ["name" =: "Dodgers", "league" =: "American", "_id" =: _id]
+        , ["name" =: "Indians", "league" =: "American"]
+        ]
   return ()
 
 insertUsers :: IO ()
-insertUsers = db $
-  insertAll_ "users" [ ["_id" =: "jane", "joined" =: parseDate "2011-03-02", "likes" =: ["golf", "racquetball"]]
-                     , ["_id" =: "joe", "joined" =: parseDate "2012-07-02", "likes" =: ["tennis", "golf", "swimming"]]
-                     , ["_id" =: "jill", "joined" =: parseDate "2013-11-17", "likes" =: ["cricket", "golf"]]
-                     ]
+insertUsers =
+  db $
+    insertAll_
+      "users"
+      [ ["_id" =: "jane", "joined" =: parseDate "2011-03-02", "likes" =: ["golf", "racquetball"]]
+      , ["_id" =: "joe", "joined" =: parseDate "2012-07-02", "likes" =: ["tennis", "golf", "swimming"]]
+      , ["_id" =: "jill", "joined" =: parseDate "2013-11-17", "likes" =: ["cricket", "golf"]]
+      ]
 
 pendingIfMongoVersion :: ((Integer, Integer) -> Bool) -> SpecWith () -> Spec
 pendingIfMongoVersion invalidVersion = before $ do
   version <- db $ extractVersion . T.splitOn "." . at "version" <$> runCommand1 "buildinfo"
   when (invalidVersion version) $ pendingWith "This test does not run in the current database version"
-  where
-    extractVersion (major:minor:_) = (read $ T.unpack major, read $ T.unpack minor)
-    extractVersion _               = error "Invalid version specification"
+ where
+  extractVersion (major : minor : _) = (read $ T.unpack major, read $ T.unpack minor)
+  extractVersion _ = error "Invalid version specification"
 
 bigDocument :: Document
-bigDocument = (flip map) [1..10000] $ \i -> (fromString $ "team" ++ (show i)) =: ("team " ++ (show i) ++ " name")
+bigDocument = (flip map) [1 .. 10000] $ \i -> (fromString $ "team" ++ (show i)) =: ("team " ++ (show i) ++ " name")
 
 fineGrainedBigDocument :: Document
-fineGrainedBigDocument = (flip map) [1..1000] $ \i -> (fromString $ "team" ++ (show i)) =: ("team " ++ (show i) ++ " name")
+fineGrainedBigDocument = (flip map) [1 .. 1000] $ \i -> (fromString $ "team" ++ (show i)) =: ("team " ++ (show i) ++ " name")
 
 hugeDocument :: Document
-hugeDocument = (flip map) [1..1000000] $ \i -> (fromString $ "team" ++ (show i)) =: ("team " ++ (show i) ++ " name")
+hugeDocument = (flip map) [1 .. 1000000] $ \i -> (fromString $ "team" ++ (show i)) =: ("team " ++ (show i) ++ " name")
+
+testUser :: (T.Text, T.Text)
+testUser = ("test_user", "123")
+
+createTestUser :: (MonadIO m) => Action m ()
+createTestUser =
+  let (username, password) = testUser
+   in addUser False username password
+
+authAsTestUser :: (MonadIO m) => Action m Bool
```

Comment:
Mongo 4.0 fails this test as well.

Reply:
Thanks, will fix ASAP

Reply:
Okay so actually the mongo server needs some configuration ahead of the tests per se to support authentication. Since authentication is on server-wide however, I think it requires all queries to use it. So I am proposing to have authentication always on in tests. See my latest commit. This allows me to deprecate all tests that depend on`authAsTestUser`.